design-documents/allowed-http-methods-for-actions.md

```diff
@@ -0,0 +1,56 @@
+### Overview
+
+To improve security and logistics we need to allow limiting Actions to processing only requests with certain HTTP methods and add those limitations to as many existing Actions as possible.
+There are many vulnerabilities caused by actions processing both GET and POST requests and thus allowing bypassing security validations like form key validation. Also limiting actions to processing only requests with certain methods would serve as self-documentation for Action classes and improve consistency of server side for client code and functional tests.
+
+### Theory of Operation
+
+Since we use by-convention routing it would be inconsistent to defined actions’ allowed HTTP methods in a config file, so it’s better to create marker interfaces like these:
+
+>interface HttpGetActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPostActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPutActionInterface extends ActionInterface
+>
+>{
+>
+>}
+
+etc.
+
+One interface per each HTTP method (using list of HTTP methods from \Zend\Http\Request::METHOD_* constants). We will have an exentable map *interface* => *HTTP method* passed to the validator to allow custom HTTP methods to be processed.
+
+Request method will be validated in the FrontController before executing a matched action, somewhat like this:
+
+>foreach ($interfaceMap as $interface => $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;if ($actionInstance instanceof $interface && $request->getMethod() !== $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;throw new NotFoundException(__(‘Page not found.’));
```

Comment:
I remember that we agreed to return 405 response code or no?

Reply:
No, we chose 404 because it's more consistent with what we do now and server must never return a 405 to GET and HEAD requests (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405)